Add passwordless login/registration to auth and web

codingllama opened the pull request:
Wire passwordless registration and authorization into Auth and Proxy APIs, thus making passwordless logins possible.

API changes are described by [RFD 52: Passwordless](https://github.com/gravitational/teleport/blob/master/rfd/0052-passwordless.md#authentication-api-changes).

#9160

codingllama committed:
Add passwordless settings to Auth protos

codingllama committed:
Update generated protos

codingllama committed:
Register: Apply DeviceUsage in lib/auth

codingllama committed:
Register: Apply DeviceUsage in lib/web

codingllama committed:
Login: Generate passwordless challenge

codingllama committed:
Login: Allow passwordless authentication

codingllama committed:
Wire passwordless in lib/web endpoints

codingllama committed:
Make mocku2f passwordless setup a bit nicer

codingllama commented on lib/auth/auth_login_test.go:
```diff
@@ -156,93 +158,15 @@ func TestServer_CreateAuthenticateChallenge_authPreference(t *testing.T) {
-				require.NoError(t, err)
-
-				// Solve challenge (client-side)
-				resp, err := test.solveChallenge(challenge)
-				authReq := AuthenticateUserRequest{
-					Username: username,
-				}
-				require.NoError(t, err)
-
-				switch {
-				case resp.GetWebauthn() != nil:
-					authReq.Webauthn = wanlib.CredentialAssertionResponseFromProto(resp.GetWebauthn())
-				case resp.GetTOTP() != nil:
-					authReq.OTP = &OTPCreds{
-						Password: []byte(password),
-						Token:    resp.GetTOTP().Code,
-					}
-				default:
-					t.Fatalf("Unexpected solved challenge type: %T", resp.Response)
-				}
-
-				// 2nd step: finish login - either SSH or Web
-				require.NoError(t, authenticate(authServer, authReq))
-			}
-		}
-		t.Run(test.name+"/ssh", makeRun(func(s *Server, req AuthenticateUserRequest) error {
-			_, err := s.AuthenticateSSHUser(AuthenticateSSHRequest{
-				AuthenticateUserRequest: req,
-				PublicKey:               []byte(sshPubKey),
-				TTL:                     24 * time.Hour,
-			})
-			return err
-		}))
-		t.Run(test.name+"/web", makeRun(func(s *Server, req AuthenticateUserRequest) error {
-			_, err := s.AuthenticateWebUser(req)
-			return err
-		}))
-	}
-}
-
```
Note: moved closer to other authentication tests.

codingllama commented on lib/auth/auth_login_test.go:
```diff
@@ -476,6 +400,310 @@ func TestCreateRegisterChallenge(t *testing.T) {
+				})
+				require.NoError(t, err)
+
+				// Solve challenge (client-side)
+				resp, err := test.solveChallenge(challenge)
+				authReq := AuthenticateUserRequest{
+					Username: username,
+				}
+				require.NoError(t, err)
+
+				switch {
+				case resp.GetWebauthn() != nil:
+					authReq.Webauthn = wanlib.CredentialAssertionResponseFromProto(resp.GetWebauthn())
+				case resp.GetTOTP() != nil:
+					authReq.OTP = &OTPCreds{
+						Password: []byte(password),
+						Token:    resp.GetTOTP().Code,
+					}
+				default:
+					t.Fatalf("Unexpected solved challenge type: %T", resp.Response)
+				}
+
+				// 2nd step: finish login - either SSH or Web
+				require.NoError(t, authenticate(authServer, authReq))
+			}
+		}
+		t.Run(test.name+"/ssh", makeRun(func(s *Server, req AuthenticateUserRequest) error {
+			_, err := s.AuthenticateSSHUser(AuthenticateSSHRequest{
+				AuthenticateUserRequest: req,
+				PublicKey:               []byte(sshPubKey),
+				TTL:                     24 * time.Hour,
+			})
+			return err
+		}))
+		t.Run(test.name+"/web", makeRun(func(s *Server, req AuthenticateUserRequest) error {
+			_, err := s.AuthenticateWebUser(req)
+			return err
+		}))
+	}
+}
```
Note: this block was moved.

codingllama commented:
Note to reviews: only you discount proto changes and code I moved around, it should be a reasonably-sized PR.

ravicious reviewed (approved):
Going through the RFD once again made reviewing this PR much easier. 👍 

There are some failing tests, but I think those are one of the flaky tests.

codingllama commented on lib/auth/auth.go:
```diff
@@ -3215,16 +3219,42 @@ func (a *Server) mfaAuthChallenge(ctx context.Context, user string) (*proto.MFAA
+		}, nil
+	}
+
+	// User required for non-passwordless.
+	if user == "" {
+		return nil, trace.BadParameter("user required")
+	}
```
It's both an attempt to be more defensive and to be explicit about when we require an username. I'm usually wary of over-validation, but I think it's reasonable for mfaAuthChallenge to care about the username after passwordless.

codingllama commented:
@ravicious 

> Going through the RFD once again made reviewing this PR much easier. 👍

Awesome, thanks for the extra effort in going through it!

> There are some failing tests, but I think those are one of the flaky tests.

Yep, it looks like the BPF tests. I think folks are working on it (if not solved already).

codingllama commented:
Friendly ping @r0mant ?

codingllama commented:
Friendly ping @r0mant ?

r0mant commented on api/client/proto/authservice.proto:
```diff
@@ -1389,6 +1410,11 @@ message CreateAuthenticateChallengeRequest {
         // VerifyAccountRecovery (step 2 of the recovery process after RPC StartAccountRecovery).
         string RecoveryStartTokenID = 2
             [ (gogoproto.jsontag) = "recovery_start_token_id,omitempty" ];
+        // Issues a challenge for the currently-authenticated user.
```
Nit: Not a huge deal but since these turn into Go fields, these should be Godoc-style comments e.g. "ContextUser is...", "Passwordless is ...", etc throughout this proto file.

codingllama replied:
Done.

codingllama replied:
Because now both messages and fields under CreateAuthenticateChallengeRequest use the same naming convention (UpperCamel), it created a naming conflict between the ContextUser/Passwordless message/fields. To solve it I moved the ContextUser and Passwordless inner messages to the top-level, which seems to be a reasonable solution. (We could also rename the message in the future without any compatibility side-effects.)

Let me know if you have other suggestions.

r0mant commented on lib/web/apiserver.go:
```diff
@@ -1723,10 +1735,14 @@ func (h *Handler) mfaLoginFinishSession(w http.ResponseWriter, r *http.Request,
-	if err := SetSessionCookie(w, req.User, session.GetName()); err != nil {
+
+	// Fetch user from session, user is empty for passwordless requests.
+	user := session.GetUser()
```
For non-passwordless, should this get the user from `req.User` as it was before?

r0mant commented on lib/auth/methods.go:
```diff
@@ -186,37 +213,62 @@ func (s *Server) authenticateUser(ctx context.Context, req AuthenticateUserReque
+	}
+	dev, user, err := s.validateMFAAuthResponse(ctx, mfaResponse, "", true /* passwordless */)
+	if err != nil {
+		log.Debugf("Passwordless authentication failed: %v", err)
```
Should this be logged at a Warn or Err level since it seems to be indicating an error? Same for the log entry below.

codingllama replied:
There are valid reasons for authentication to fail, it doesn't necessarily mean a noticeable event (or bug). This is akin to the [debug statement on authenticateUser](https://github.com/gravitational/teleport/blob/61fce15ee218a29c425ba4082b37e81ab0dcdec3/lib/auth/methods.go#L153).

Kept.

r0mant commented on lib/auth/methods.go:
```diff
@@ -105,14 +118,27 @@ func (s *Server) AuthenticateUser(req AuthenticateUserRequest) error {
 		log.WithError(err).Warn("Failed to emit login event.")
 	}
-	return err
+	return user, err
```
```suggestion
	return user, trace.Wrap(err)
```

codingllama replied:
Done.

r0mant reviewed (approved):
A few nits but nothing major.

codingllama commented on api/client/proto/authservice.proto:
```diff
@@ -1412,6 +1438,9 @@ message CreateRegisterChallengeRequest {
     DeviceType DeviceType = 2 [ (gogoproto.jsontag) = "device_type" ];
+    // Requested usage for the device.
+    // Defaults to DEVICE_USAGE_MFA.
+    DeviceUsage device_usage = 3 [ (gogoproto.jsontag) = "device_usage,omitempty" ];
```
Done. I'd love to get us into a consistent proto style, but this is out of scope for here.

codingllama committed:
Use consistent proto naming and comments

Using types and field names with the same casing strategy causes a type
name clash on CreateAuthenticateChallengeRequest, hence the messages
being moved outside of it.

codingllama commented on lib/client/weblogin.go:
```diff
@@ -87,6 +87,9 @@ type SSOLoginConsoleResponse struct {
 type MFAChallengeRequest struct {
 	User string `json:"user"`
 	Pass string `json:"pass"`
+	// Passwordless explicitly request a challenge for passwordless/usernameless
```
Done.

codingllama committed:
Tweak comments and trace.Wrap

codingllama committed:
Update generated protos

codingllama committed:
Update proto usage according to new message names

codingllama commented:
PTAL?
